src/style/mod.rs

```diff
@@ -239,10 +252,14 @@ impl<F: MonoFont> TabSize<F> {
 /// [`TextBoxStyleBuilder`]: builder/struct.TextBoxStyleBuilder.html
 /// [`new`]: #method.new
 /// [`from_text_style`]: #method.from_text_style
-#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
-pub struct TextBoxStyle<C, F, A, V, H> {
+#[derive(Copy, Clone, Debug)]
+pub struct TextBoxStyle<C, F, A, V, H>
```

Comment:
Should we add a background color attribute to this? IMO paragraph and character background should be separate.

Reply:
Ugh... Normally I'd place my vote for yes, but I didn't really want to implement a box-wide background.

Reply:
We don't need to implement this right now, but we might have to consider this in the text renderer API. Rendering text without any transparency can use the more efficient `DrawTarget::fill_contiguous`. But at the moment e-t couldn't use this code path if the character background is set to transparent and the box background is set to a color.

Reply:
I don't think we can draw a whole textbox in one go anyway, unless the text renderers provide pixel iterators (and this PR transitions away from exactly that). Normally, the background would be e-t's job, since it's the one who knows about where a line begins and ends.

e-t can draw the background, but how should we handle the case where the character style is transparent background, but the box has a solid color?

Reply:
In the simplest implementation of this feature e-t would fill the background of the box with a solid color and then draw text with a transparent character background onto it.

If we would extend the text renderer API to allow the caller to override the character background color, we could implement more efficient drawing. e-t could draw each line using a sequence of `draw_text` and `draw_whitespace` calls. The empty space on the left and right of each line and the spacing between lines could then be filled using `DrawTarget::fill_solid`. While this is still a relatively large number of draw calls it should still be better than using `draw_iter` to render transparent text (at least if the display driver doesn't implement caching).

Reply:
Changing the character style is already a must to get escape sequences working.

Reply:
I hadn't considered that. So we need a way to generically change colors for all character styles or a way to override the colors using the text renderer API.

Reply:
That or e-t could provide the required abstraction like in the impl in this pr. It's not only the color but also the decoration and in the future perhaps the font itself as well.

Reply:
Here I'd  probably require implementing a trait that has an `fn apply_sgr(Sgr)` function.

Reply:
TBH I haven't looked at the PR in detail yet. But IMO it would be nice if e-t could work with any font renderer without adding special support for it in e-t.

Reply:
I wish specialization was stable. That way we could make escape sequences optional if the renderer implements the required trait, but e-t would otherwise work with any renderer.

Reply:
I would like to keep SGR attributes out of the text renderers. We could add a trait that can be used to modify character styles. Implementations of this trait would ignore any call that isn't supported by the renderer.

The trait would something like this:
```rust
trait CharacterStyle<C> {
    fn set_text_color(&mut self, color: Option<C>);
    fn set_background_color(&mut self, color: Option<C>);
    fn set_underline_color(&mut self, color: Option<C>);
    fn set_strikethrough(&mut self, color: Option<C>);
    fn set_italic(&mut self, italic: bool);
    fn set_bold(&mut self, bold: bool);
}
```

Reply:
> I wish specialization was stable. That way we could make escape sequences optional if the renderer implements the required trait, but e-t would otherwise work with any renderer.

Specialization would be a really useful feature, that I've also missed several times.

Reply:
That would work. If someone wants to do something fancier, I think e-t would still provide an `SrgStylable` or similar that blanket implements for `T: CharacterStyle`.

Bold? Italic? That's a lot of bitmaps for `MonoTextStyle` ;)

Reply:
> Bold? Italic? That's a lot of bitmaps for `MonoTextStyle` ;)

These settings weren't targeting `MonoTextStyle` ;). I might add some type of font family support to the BDF renderer, where this would become useful.